src/main/scala/viper/gobra/frontend/info/base/Type.scala

```diff
@@ -210,4 +217,59 @@ object Type {
     * vector storing the receiver's type for methods and mpredicates.
     */
   case class AbstractType(messages: (PNode, Vector[Type]) => Messages, typing: Vector[Type] ==> FunctionT) extends PrettyType("abstract")
+
+  trait TypeNode extends Node {
+    /**
+      * Applies the type parameter substitution f to this type
+      */
+    def substitute(f: PartialFunction[PIdnDef, Type]): this.type = {
+      this.transform({
+        case TypeParameterT(id, _, _) if f.isDefinedAt(id) => f(id)
+      })
+    }
+
+    /**
+      * Returns uninstantiated in the type
+      */
+    def uninstantiatedTypeParameters(symb: SymbolTable.WithTypeParameters): Seq[PIdnDef] = {
+      /**
+        * collect all type parameters in the type and filter out type parameters that are not declared by the symb
+        * (we consider type parameters that are not declared by the symbol as instantiated)
+        */
+      this.deepCollect({
+        case t: TypeParameterT => t.id
+      }).intersect(symb.typeParameters.map(_.id))
+    }
+
+    override def info: Parser.Info = Source.Parser.Unsourced
```

Comment:
Why is this necessary?

Reply:
viper.gobra.ast.internal.info requires this to be defined

Reply:
I just noticed that TypeNode extends the internal Node. This seems very confusing, since this step is before the internal ast, so why is it extending the internal ast node?

Reply:
We discussed it like this in a meeting. The idea was that we don't have to implement the logic for transform and deepCollect twice.